pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,242 @@
+	log.VEventf(ctx, 2, "resetting sink to span %s", newSpan)
+	if err := s.maybeFlushNonExtendableSpans(newSpan); err != nil {
+		return err
+	}
+	if s.out == nil {
+		if err := s.open(ctx); err != nil {
+			return err
+		}
+	}
+	// At this point, we can assume the new span is either contiguous with the
+	// last span or comes after the last span (if it exists).
+	var lastFile *backuppb.BackupManifest_File
+	if len(s.flushedFiles) > 0 {
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	if lastFile != nil {
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
+		if isContiguousSpan(lastFile.Span, newSpan) &&
+			(s.midRow || lastFile.EntryCounts.DataSize < fileSpanByteLimit) {
+			log.VEventf(ctx, 2, "extending span %s to %s", lastFile.Span, newSpan)
+			s.stats.spanGrows++
+			lastFile.Span.EndKey = newSpan.EndKey
+			return nil
+		}
+	}
+
+	s.flushedFiles = append(
+		s.flushedFiles,
+		backuppb.BackupManifest_File{
+			Span: newSpan,
+			Path: s.outName,
+		},
+	)
+	s.prevKey = nil
+	return nil
+}
+
+func (s *SSTSinkKeyWriter) Flush(ctx context.Context) error {
+	if len(s.flushedFiles) > 0 {
+		lastFile := s.flushedFiles[len(s.flushedFiles)-1]
+		// If WriteKey was used, it is possible that the last written key was
+		// mid-row, but because no keys were written after, s.midRow was not updated.
+		// To ensure this, we update midRow using the end of the span that the last
+		// key belonged to.
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
+	}
+	err := s.FileSSTSink.Flush(ctx)
+	return err
+}
+
+// maybeFlushNonExtendableSpans checks if the new span's startKey precedes the
+// last spans' endKey or if the two spans do not share the same prefix based on
+// the sink's ElideMode. If either of these conditions are met, the current
+// backup file is flushed and a new one is created.
+func (s *SSTSinkKeyWriter) maybeFlushNonExtendableSpans(newSpan roachpb.Span) error {
+	if len(s.flushedFiles) == 0 {
+		return nil
+	}
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
+	samePrefix, err := sameElidedPrefix(newSpan.Key, lastFile.Span.EndKey, s.conf.ElideMode)
+	if err != nil {
+		return err
+	}
+	if !samePrefix || newSpan.Key.Compare(lastFile.Span.EndKey) < 0 {
+		log.VEventf(
+			s.ctx, 1, "flushing backup file %s of size %d because new span %s is not contiguous with last span %s",
+			s.outName, s.flushedSize, newSpan, lastFile.Span,
+		)
+		s.stats.oooFlushes++
+		if err := s.Flush(s.ctx); err != nil {
+			return err
+		}
+		return nil
+	}
+	return nil
+}
+
+// maybeDoSizeFlush checks if either a BackupManifest_File needs to be flushed
+// due to reaching the fileSpanByteLimit or if the entire SST needs to be flushed
+// due to reaching targetFileSize. If either type of flush is performed, a new
+// BackupManifest_File is created for the nextKey, so that is returned.
+func (s *SSTSinkKeyWriter) maybeDoSizeFlush(
+	ctx context.Context, nextKey roachpb.Key,
+) (*backuppb.BackupManifest_File, error) {
```

Comment:
nit: it's best practice to avoid returning an object that is already accessible to the caller. Here, the caller should assume that they only ever modify `s.flushedFiles[len(s.flushedFiles)-1]`. So, i think this func should return err

Reply:
Sounds good, I'll keep that in mind